[loop_tool] Add integration and tests

bwasti opened the pull request:
Update: it's working :)

```
python -m compiler_gym.bin.random_search --env=looptool-v0 --benchmark=loop_tool-v0/8388608 --reward=flops --nproc=1 --runtime=60
```


This PR scaffolds integration of the ~(nearly open sourced)~ loop_tool.  ~This doesn't include the full integration (work in progress).~

I wanted to put this up to get feedback on the proposed integration points.  A couple questions

1. should this live somewhere else? envs/loop_tool seems reasonable, but the changes required to the build and requirements.txt seem somewhat invasive
2. if this should live in envs/loop_tool, should I try to refactor things to make loop_tool optional? (basically just remove from requirements.txt and add a fallback error message asking folks to `pip install loop_tool`)
3. Is CompilerSession really an env? I think I might be getting semantics wrong here, but I'm not sure the exact parallels to llvm's integration

bwasti committed:
[loop_tool] Add scaffolded integration

bwasti commented on setup.py:
```diff
@@ -54,7 +54,7 @@ def get_tag(self):
         "compiler_gym.envs.llvm.service.passes",
         "compiler_gym.envs.llvm.service",
         "compiler_gym.envs.llvm",
-        "compiler_gym.envs.llvm",
```
I think this was an accidental redundancy :) (I know it looks like I removed llvm, but line 56 still imports it)

ChrisCummins commented on tests/loop_tool/actions.py:
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Please name test files with a `_test.py` suffix so that they are run by the pytest collector on `make install-test`

ChrisCummins commented on compiler_gym/envs/loop_tool/loop_tool_env.py:
```diff
@@ -0,0 +1,83 @@
+            name="flops",
+            scalar_double_range=ScalarRange(),
+            deterministic=False,
+            platform_dependent=False,
```
```suggestion
            platform_dependent=True,
```

i think

ChrisCummins reviewed (commented):
Aye up, I know that this is still WIP but I thought I'd take a sneaky peek :)

> should this live somewhere else? envs/loop_tool seems reasonable, but the changes required to the build and requirements.txt seem somewhat invasive

Hard to say without seeing the changes, but I think envs/loop_tool is the right place. I feel like `pip install compiler_gym` should ship with all of the environments. In the case of env-specific pip requirements, you could always add a seperate `compiler_gym/envs/loop_tool/requirements.txt` file and `-r` import it from `compiler_gym/requirements.txt` to keep it clear what the deps are needed for. [example](https://github.com/facebookresearch/CompilerGym/blob/development/requirements.txt)

> if this should live in envs/loop_tool, should I try to refactor things to make loop_tool optional? (basically just remove from requirements.txt and add a fallback error message asking folks to pip install loop_tool)

I don't think that'll be necessary.

> Is CompilerSession really an env? I think I might be getting semantics wrong here, but I'm not sure the exact parallels to llvm's integration

Not quite. An env is the *frontend* python object that you interact with. The base class for these is [CompilerEnv](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/compiler_env.py#L70-L129). A [CompilationSession](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/service/compilation_session.py#L17-L36) is the *backend* object that represents a single episode, e.g.

```py
env = gym.make(...)  # A CompilerEnv is born
env.reset()  # starts a CompilationSession in the backend
# ... env.step() etc
env.reset()  # starts a new CompilationSession in the backend
```

[Here is the CompilationSession for LLVM](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/llvm/service/LlvmSession.h#L31-L38). There is also a custom [CompilerEnv for LLVM](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/llvm/llvm_env.py#L75-L91).

Lemme know if that makes sense, there's a very good chance to the docs could be improved to make things more clear :)

Commit:
[loop_tool] Use dev branch for loop-tool-py dependency.

Use the development repo for loop-tool-py dependency. Once pushed to
PyPi this commit can be dropped.

Commit:
[loop_tool] Integrate loop tool scaffolding into build system.

Commit:
[loop_tool] Set verbose logging for loop_tool test.

Once the tests pass this commit can be dropped.

Commit:
[loop_tool] Add dummy values for observation space.

To fix failing //tests/bin:service_bin_test test.

ChrisCummins commented:
I have a small patchset of changes for this branch:

1. Added a placeholder LoopToolEnv class in case you want to add any custom behavior for loop tool.
2. Made the loop tool env compatible with bazel / pip install.
3. Moved the service into a "service" subpackage (just for consistency with llvm, maybe this is overkill).

There is one failing test:

```
FAILED tests/loop_tool/actions_test.py::test_basic - compiler_gym.service.connection.SessionNotFound: Must call reset() before step()
```

3 of my 5 commits are just for dev / debugging and can be dropped prior to merging. Check commit messages.

Cheers,
Chris

bwasti committed:
fix error and add loop_tree observation space

bwasti committed:
fix requirements to use PyPI loop_tool package

bwasti committed:
add more tests

ChrisCummins commented:
Would you mind pulling and rebasing on the latest `development`? A fix which landed yesterday (#315) is needed to run the CI tests.

Cheers,
Chris

ChrisCummins commented on compiler_gym/envs/loop_tool/service/loop_tool_compilation_session.py:
```diff
@@ -0,0 +1,272 @@
+        ObservationSpace(
+            name="action_state",
+            int64_range_list=ScalarRangeList(
+                # FIXME(bwasti): Dummy values.
```
Is this fixed now?

ChrisCummins commented on compiler_gym/envs/loop_tool/service/loop_tool_compilation_session.py:
```diff
@@ -0,0 +1,272 @@
+            platform_dependent=False,
+            default_value=Observation(
+                int64_list=Int64List(
+                    # FIXME(bwasti): Dummy values.
```
and this one

ChrisCummins commented on compiler_gym/envs/loop_tool/service/loop_tool_compilation_session.py:
```diff
@@ -0,0 +1,272 @@
+        B.set(self.Bp)
+        iters = 10000
+        # warmup
+        for i in range(50):
```
Should `50` be a constant variable for clarity?

ChrisCummins commented on compiler_gym/envs/loop_tool/service/loop_tool_compilation_session.py:
```diff
@@ -0,0 +1,272 @@
+        return flops
+
+    def get_observation(self, observation_space: ObservationSpace) -> Observation:
+        # TODO populate
```
Is this TODO done?

ChrisCummins commented on compiler_gym/requirements.txt:
```diff
@@ -4,6 +4,7 @@ fasteners>=0.15
 grpcio>=1.32.0
 gym>=0.18.0
 humanize>=2.6.0
+loop_tool_py=0.0.4
```
```suggestion
loop_tool_py==0.0.4
```

ChrisCummins commented on compiler_gym/envs/loop_tool/service/loop_tool_compilation_session.py:
```diff
@@ -0,0 +1,272 @@
+class LoopToolCompilationSession(CompilationSession):
+    """Represents an instance of an interactive loop_tool session."""
+
+    compiler_version: str = "0.0.1"
```
I think this might be out of date :) Perhaps something like this?
```suggestion
    compiler_version: str = pkg_resources.get_distribution('loop-tool-py').version
```

ChrisCummins reviewed (commented):
This looks great, @bwasti. I'm going to work through the loop_tool tutorial on Monday morning so that I have a better understand of what the environment is doing. So in the meantime, here's a bunch of tiny nitpicks from a read through of the code :-)

Cheers,
Chris

bwasti committed:
address comments on PR

bwasti committed:
fix test with new auto-unroll logic in search space

bwasti committed:
add documentation

ChrisCummins commented on Makefile:
```diff
@@ -62,7 +62,8 @@ Documentation
         Build the HTML documentation and serve them on localhost:8000. Changes
         to the documentation will automatically trigger incremental rebuilds
-        and reload the changes.
+        and reload the changes.  To change the host and port, set the SPHINXOPTS
```
Nice, thanks for this!

ChrisCummins commented on setup.py:
```diff
@@ -18,6 +18,9 @@
         long_description = f.read()
 with open("compiler_gym/requirements.txt") as f:
     requirements = [ln.split("#")[0].rstrip() for ln in f.readlines()]
+    # Split the PyPi packages from dependency links.
+    dependency_links = [ln for ln in requirements if ln.startswith("git+")]
+    requirements = [ln for ln in requirements if not ln.startswith("git+")]
```
Is this still needed?

ChrisCummins commented on setup.py:
```diff
@@ -92,6 +97,7 @@ def get_tag(self):
         ]
     },
     install_requires=requirements,
+    dependency_links=dependency_links,
```
Is this still needed?

ChrisCummins commented on tests/loop_tool/actions_test.py:
```diff
@@ -0,0 +1,141 @@
+
+
+def test_basic():
+    env = compiler_gym.make("looptool-v0")
```
Could you refactor this (and other `test_xxx()` funcs) to use the pattern:

```py
with compiler_gym.make("looptool-v0") as env:
    # test body
```

Using the `with` syntax ensures that `env.close()` is called so that no compiler gym service is orphaned.

ChrisCummins commented on tests/loop_tool/actions_test.py:
```diff
@@ -0,0 +1,141 @@
+
+
+if __name__ == "__main__":
+    main(debug_level=4)
```
```suggestion
    main()
```

ChrisCummins reviewed (commented):
This is looking really good! Thanks a lot for the extra docs page.

Does the loop_tool environment require cuda to run? Could you add an "Installation" section (e.g. [for GCC](https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/envs/gcc.rst#installation)) which describes what, if anything, users need to do to use the loop_tool env? And then at runtime raise an `EnvironmentNotSupported` error if those runtime requirements are not met (e.g. [for GCC](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/gcc/gcc.py#L223-L231)). At the moment the env initialization fails with:

```
                    raise ServiceError(
                        f"RPC call returned status code {e.code()} and error `{e.details()}`"
>                   ) from None
E                   compiler_gym.service.connection.ServiceError: RPC call returned status code StatusCode.INTERNAL and error `assertion: set failed @ /tmp/pip-req-build-z726l9ei/src/frontends/python.cpp:49 cannot find hardware: cuda`
```

Cheers,
Chris

bwasti committed:
add CPU fallback

bwasti committed:
fix up w.r.t PR comments

ChrisCummins committed:
[docs] Small formatting tweaks for loop_tool reference.

ChrisCummins committed:
[loop_tool] Set environment name to "loop_tool-v0".

ChrisCummins committed:
[loop_tool] Set default reward and observation spaces.

ChrisCummins committed:
[ci] Run loop_tool tests as separate job.

ChrisCummins commented:
LGTM. Merging, thanks a lot @bwasti for adding a whole new optimization domain to CompilerGym!

Cheers,
Chris
